executables/bal.go

```diff
@@ -0,0 +1,306 @@
+	if stat, err := os.Stat(javaPath); err == nil && stat.IsDir() {
+		javaHome := javaPath
+		os.Setenv("JAVA_HOME", javaHome)
+	}
+
+	switch runtime.GOOS {
+	case "darwin":
+
+		javaHome := os.Getenv("JAVA_HOME")
+		javaVersion := os.Getenv("JAVA_VERSION")
+		if javaHome == "" {
+			if javaVersion == "" {
+				javaHomeCmd, _ := exec.Command("/usr/libexec/java_home").Output()
+				javaHome = string(javaHomeCmd)
+				os.Setenv("JAVA_HOME", javaHome)
+			} else {
+				// If JAVA_HOME is not set, but JAVA_VERSION is set, get the Java home for the specified version
+				javaHomeCmd := exec.Command("/usr/libexec/java_home", "-v", javaVersion)
+				javaHomeOutput, err := javaHomeCmd.Output()
+				if err == nil {
+					javaHome = string(javaHomeOutput)
+					os.Setenv("JAVA_HOME", javaHome)
+				} else {
+					fmt.Printf("Error determining Java home for version %s: %v\n", javaVersion, err)
+					os.Exit(1)
+				}
+			}
+		}
+	}
+
+	prg, err := os.Executable()
+	//prg = "/usr/lib/ballerina/distributions/ballerina-2201.8.4/bin/bal"
+	if err != nil {
+		fmt.Println("Error getting executable path:", err)
+		os.Exit(1)
+	}
+
+	for {
+		link, err := filepath.EvalSymlinks(prg)
+		if err != nil {
+			fmt.Println("Error resolving symbolic link:", err)
+			os.Exit(1)
+		}
+
+		if link == prg {
+			break
+		}
+
+		prg = link
+	}
+
+	// Get the directory of the script
+	prgDir := filepath.Dir(prg)
+
+	// Set BALLERINA_HOME
+	ballerinaHome, err = filepath.Abs(filepath.Join(prgDir, ".."))
+	if err != nil {
+		fmt.Println("Error getting absolute path:", err)
+		os.Exit(1)
+	}
+	if javaCmd == "" {
+		if javaHome != "" {
+			var javacmdPath string
+			if _, err := os.Stat(filepath.Join(javaHome, "jre", "sh", "java")); err == nil {
+				javacmdPath = filepath.Join(javaHome, "jre", "sh", "java")
+			} else {
+				javacmdPath = filepath.Join(javaHome, "bin", "java")
+			}
+			if _, err := os.Stat(javacmdPath); err == nil {
+				javaCmd = javacmdPath
+			} else {
+				fmt.Printf("Error: Unable to find executable JAVACMD in specified JAVA_HOME: %s\n", javaHome)
+				os.Exit(1)
+			}
+		} else {
+			javaCmd = "java"
+		}
+	}
+
+	javaCmd = strings.TrimSpace(javaCmd)
+	var ballerinaCLIWidth string
+
+	// Check if tput is available
+	if !commandExists("tput") {
+		ballerinaCLIWidth = "80"
+	} else {
+		if cols, err := getTerminalColumns(); err != nil {
+			ballerinaCLIWidth = "80"
+		} else {
+			ballerinaCLIWidth = strconv.Itoa(cols)
+		}
+	}
+	os.Setenv("BALLERINA_CLI_WIDTH", ballerinaCLIWidth)
+
+	//Setting Ballerina debug port
+
+	balJavaDebug, isSet := os.LookupEnv("BAL_JAVA_DEBUG")
+	javaOpts := ""
+	if isSet {
+		if balJavaDebug == "" {
+			fmt.Println("Please specify the debug port for the BAL_JAVA_DEBUG variable")
+			os.Exit(1)
+		} else {
+			javaOpts = os.Getenv("JAVA_OPTS")
+
+			if javaOpts != "" {
+				fmt.Println("Warning !!! User specified JAVA_OPTS may interfere with BAL_JAVA_DEBUG")
+			}
+
+			balDebugOpts := os.Getenv("BAL_DEBUG_OPTS")
```

Comment:
Since you have accessed the environment variables throughout the program, setting them as constants at the beginning of the program would be better.